### ISSUE TITLE
KNN Mahalanobis distance error

### ISSUE THREAD

**Issue by hanshupe** [yzhao062/pyod#23]
Hi,

When I use the Mahalanobis metric for KNN I always get the error "Must provide either V or VI for Mahalanobis distance" even when I provide V with metric_params. The same request works with sklearn.neighbors.

```

from pyod.models.knn import KNN  
from pyod.utils.data import generate_data
from sklearn.neighbors import NearestNeighbors
import numpy as np

contamination = 0.1  
n_train = 200  
n_test = 100 

X_train, y_train, X_test, y_test = generate_data(n_train=n_train, n_test=n_test, contamination=contamination)

#Doesn't work (Must provide either V or VI for Mahalanobis distance)
clf = KNN(algorithm='brute', metric='mahalanobis', metric_params={'V': np.cov(X_train)})
clf.fit(X_train)

#Works
nn = NearestNeighbors(algorithm='brute', metric='mahalanobis', metric_params={'V': np.cov(X_train)})
nn.fit(X_train)
```

**Comment by yzhao062** [comment]
Thanks for reporting this. It is due to the problem of passing parameters into KDTree. 
Specifically, it is caused by the difference between the parameters of NearestNeighbors and KDTree...KDTree has a distinct way of using customized distance metrics.

I will fix it in this next few days and update you here.

**Comment by yzhao062** [comment]
Sorry for being late on this. I was exploring several options but get stuck by passing in customized metric to sklearn KD_tree. In PyOD, KNN detector uses a KD-tree internally.

I tried one solution to pass in mahalanobis distance:
> metric = DistanceMetric.get_metric('mahalanobis', V=np.cov(X_test))
> 
> #Doesn't work (Must provide either V or VI for Mahalanobis distance)
> clf = KNN(algorithm='brute', metric=metric.pairwise)
> clf.fit(X_train)
> #https://github.com/scikit-learn/scikit-learn/issues/8890

The reason is that I built a KD-tree and also a NearestNeighbors for KNN, which should be combined into one. Will provide an update later...

As an alternative, I will recommend you to check out MCD model in PyOD that uses Mahalanobis:

>     First fit a minimum covariance determinant model and then compute the
>     Mahalanobis distance as the outlier degree of the data

**Comment by hanshupe** [comment]
Hi can I ask if there will be a solution for the bug? Thanks

**Comment by evanmiller29** [comment]
@yzhao062 if you're too busy I could give it a try

**Comment by yzhao062** [comment]
@evanmiller29 that will be perfect! 
The current problem is I am using a KD tree and a NearestNeighbors at the same time...which might be consolidated into one with consistent parameters (KD tree has different parameters than NearestNeighbors which makes the fix more complicated). 
Let me know if you need any other tips & thanks for the help.

**Comment by sdoering** [comment]
Hello @evanmiller29, hello @yzhao062,

any news on this issue? Am evaluating pyod and stumbled over this issue. Some update here would be great. 

Thanks for your efforts and work. 

All the best

**Comment by yzhao062** [comment]
@sdoering I am back to fix this one. Someone brought this into my attention yesterday (https://stackoverflow.com/q/55728669/4168429?sem=2). I will aim to provide a solution this week (doing some experiments now). Stay tuned :)

**Comment by yzhao062** [comment]
The bug is fixed by commit https://github.com/yzhao062/pyod/commit/db41e8e03e3901782eb68ff77ab95fc89e82bd92

The next release V0.7.0 will include this commit. See https://github.com/yzhao062/pyod/blob/development/examples/knn_mahalanobis_example.py for example. 

Before the new version is officially released, you could overwrite the PyOD source file knn.py by https://github.com/yzhao062/pyod/blob/development/pyod/models/knn.py .

The latest commit is not yet pushed to master branch but development branch only (https://github.com/yzhao062/pyod/tree/development)

**Comment by yzhao062** [comment]
Closed by #84. Available in the latest version ([v0.7.0](https://pypi.org/project/pyod/)).

**Comment by lachhebo** [comment]
When your using the function np.cov, you need to set the parameter rowvar to False like that : 
`np.cov(X_train, rowvar =False)`

**Comment by yzhao062** [comment]
@lachhebo Yeah I think so. See the example provided here: https://github.com/yzhao062/pyod/blob/development/examples/knn_mahalanobis_example.py
